### ISSUE TITLE
2+2+1 Major usage modes for Napari

### ISSUE THREAD

**Issue by sofroniewn** [Napari/napari-core#28]
## Description
Based on conversations / discussions on other issues and PRs (like https://github.com/Napari/napari-gui/pull/126) I wanted to create an issue with some of the major potential usage modes of napari. These ideas are very much up for further debate and we might end-up somewhere quite far from this initial post.

1. Use napari viewer and napari plugins from jupyter notebooks or python scripts. These users are likely to be relatively proficient in python, understand the basics of python module imports, dependency management. These users might develop their own plugins. They want to be able to do things like  `from napari.gui import Viewer` and `from napari.plugins import ExamplePlugin` or just `import ExamplePlugin` and run `viewer.add_plugin(ExamplePlugin())` to just add this plugin to the viewer - though see https://github.com/Napari/napari-gui/pull/126 for alternative approaches. The key value add for them in the notebook is being able to interactively do data analysis on images, getting annotations etc back from the viewer to the notebook and then update the viewer from the notebook. For these users flexibility, customizability, and conformation to the norms of python are key. I expect these users to be high-value users of Napari as they are likely to build Napari plugins and evangelize the ecosystem if they have a good experience and see value in it. 

2. Use the napari application. These users may have no experience with python and have no understanding of dependency management or versioning. They will open the napari application from a desktop application and then add images and plugins from the dropdown menu. Once a plugin has been installed into the application it will be permanently installed (until removed). They are able to run update scripts to ensure they have the latest versions of the viewer and plugins, but if things break they will have no recourse to fix them. For this use case we might need to consider things like sandboxing of plugins / resolving dependency conflicts etc to ensure the app is robust to "third party" plugins, but we'll also need to ensure performance and interactivity remains high. These users may ultimately represent the majority of Napari users if we achieve high-penetrance in communities doing lots of image analysis but with less python experience (such as biology), but I think we can focus our initial development on our first use case before biting off too large an engineering challenge prematurely.

3. These users are the same as in (1 - scripts / notebooks) - but now the datasets are remote, i.e. on big gpu machines or cloud computing environments for training ML models. This use case could be a major one for us and I think is a high priority to support. It might relate to how we think about sandboxing plugins for dependency management in the application use case or how we deal with slow plugins (i.e. with long computation times that we don't want to block the gui thread) in all use cases.

4. These users are the same as in (2 - fully managed application) - but they also have datasets that are remote and want to do things like train ML models. This use case could be more complex, but also valuable.

5. Use the napari viewer embedded in their own applications. These users are likely to be developers that are highly proficient in python / PyQt. Use cases for this include building highly custom domain specific that have a need for different views and interactivity not provided by the plugin ecosystem. For this usage people will do something like `from napari.gui import Viewer` and then pretty much be on their own to add the PyQt elements into their application. I see this as a relatively low priority use case for us to support, but that could change as we get a better sense of what people want.

**Comment by oeway** [comment]
Hi, glad to see this overview of usage modes. Closely related to our discussion in thread https://github.com/Napari/napari-core/issues/27, I think there is another way which ImJoy and Napari can work together. In particular, I am thinking about we can work together to support point 3 and 4 which ImJoy is designed for. To support remote computing, we can basically run Napari as a computational backend connected with [ImJoy `window` plugins](https://imjoy.io/docs/#/development?id=window).

It is very feasible because ImJoy use conda/pip to solve the `requirements` and setup virtual environments, so we can basically add all the `requirements` of Napari and run it. To make it clear, I made a demo plugin which installs `napari-core` and popup an alert dialog in ImJoy with the napari version number: [source code](https://gist.github.com/oeway/78547b315fb65e7b6234c25b509bdfdd), [try it in ImJoy](https://imjoy.io/#/app?plugin=https://gist.githubusercontent.com/oeway/78547b315fb65e7b6234c25b509bdfdd/raw/462b9c7894369e3534a91a6c9abcfb9ff2c9b82c/Napari-version.imjoy.html).

Since the development is done in jupyter notebook or python scripts, we can then basically copy and paste those code into the ImJoy plugin file. It will already allow the user to do computational tasks with Napari and its plugins, remotely.

Now, of course, it would be much nicer to see the outputs and interact with Napari interface, this would mean that there will be two type of backend for displaying data and take user inputs, this is very much like `matplotlib` which has various backends including PyQT5 and the `%matplotlib inline` mode in jupyter notebook. I think we can basically take a similar strategy to support both PyQT mode and ImJoy `window` mode. For the latter, we can port the Napari-GUI api to a ImJoy `window` plugin and bridge it with RPC calls.

Think a bit further, we can even try to run the entire ImJoy runtime environment inside a jupyter notebook, that would means the user will be able to debug directly in a jupyter notebook with `inline` display mode. And Napari will be able to switch between `inline` mode and `pyQT` mode in a notebook. Once the development phase is done, the application can run in Napari desktop app, or remotely with ImJoy.

**Comment by sofroniewn** [comment]
> To support remote computing, we can basically run Napari as a computational backend connected with [ImJoy `window` plugins](https://imjoy.io/docs/#/development?id=window).

Given the goal of Napari plugins as being fairly lightweight (either simply annotated python, or possibly including some lightweight wrapping for highly customized plugins) I think at this point you'd pretty much just have python as a backend and be using imjoy (which is fine!).

I think the value add of Napari will be to use Napari viewer and interact with the plugins inside it, but exactly how and where we want to do the remoting is still to be decided.
